### PR TITLE
feat: video-listing page

### PR DESCRIPTION
**Pull request by bhtibrewal**
done :
- product listing
- add to like, watch later, history

pages to check:
- [BasicCard.jsx](https://github.com/bhtibrewal/artsy-stream/compare/video-listing?expand=1#diff-2920253d0f2f4638587cb81fb8f91233da26d612a2e1680ec58050334d124219)
- [filter context ](https://github.com/bhtibrewal/artsy-stream/compare/video-listing?expand=1#diff-1944caff37499e9114cfd1af699915face8b0e8b0cf3f1a45d63f5b4f48e306d)
- [video context](https://github.com/bhtibrewal/artsy-stream/compare/video-listing?expand=1#diff-1000565f2b309536583683a2c63f6950485ff7d8a9eb987be695c3e3c9dcbb32)

### CODE REVIEW

File: src/components/card/BasicCard.jsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    } else navigate("/sign-in");
+  };
+  return (
+    <div className="card w-30 basic">
+      <div className="img-sec">
+        <span className="card-tag">{tag}</span>
+        <img className="card-img" src={img_src} alt={`${title} ${creator}`} />
+        <button
+          className="btn play-btn"
+          onClick={() => openVideoAndHandleHistory()}
```
Comment (Naman-Saxena1) [inline src/components/card/BasicCard.jsx:62]: I really liked you used descriptive names everywhere, makes easier for other person  to understand code.

File: src/context/VideoFilterContext.jsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  );
+};
+const useVideoFilter = () => {
+  const context = useContext(VideoFilterContext);
+  if (!context) {
+    throw new Error(
+      `use useVideoFilter must be used inside a context provider`
+    );
+  }
+  return context;
```
Comment (Naman-Saxena1) [inline src/context/VideoFilterContext.jsx:36]: Interesting, I never thought about handling context related errors. Though in React I think we can also use Error Boundaries. But this is great, at least you thought about it.

**Review by Naman-Saxena1 — APPROVED** [review]
## Code looks great, nice work!
## I have 1 UI suggestion, can add some style to active tab in sidebar. Everything else is great 👍 
Capture5

File: src/components/card/BasicCard.jsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      <div className="content">
+        <div className="card-header">
+          <h1>{title}</h1>
+          <h2>by {creator}</h2>
+          <Rating rating={rating} />
+          <div className="video-actions">
+            <div
+              className=" video_card-btn center"
+              title="Like"
+              onClick={() => handleLike()}
```
Comment (Ankur9669) [inline src/components/card/BasicCard.jsx:76]: Suggestion: You can simply write handleLike here

File: src/context/VideoFilterContext.jsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { useReducer, createContext, useContext } from "react";
+
+const VideoFilterContext = createContext();
+
+const VideoFilterProvider = ({ children }) => {
+  const initialFilterState = {
+    showCategories: [],
+  };
+  const filter_reducer_fn = (state, { type, payload }) => {
```
Comment (Ankur9669) [inline src/context/VideoFilterContext.jsx:9]: Suggestion: As you are using camelCasing everywhere why have you used snakeCase here?

File: src/context/VideoStateContext.jsx [hunk 5]
```diff
@@ -11,22 +12,51 @@ const VideoStateProvider = ({ children }) => {
...
+      case "HANDLE_LIKES":
+        return { ...state, likedList: [...payload] };
+      case "HANDLE_HISTORY":
+        return { ...state, history: [...payload] };
+      case "HANDLE_WATCHLATER":
+        return { ...state, watchLater: [...payload] };
+      case "HANDLE_PLAYLISTS":
+        return { ...state, playLists: [...payload] };
+      case "HANDLE_PLAYLIST":
+        console.log(payload);
```
Comment (Ankur9669) [inline src/context/VideoStateContext.jsx:39]: Can remove console statement

File: src/context/VideoStateContext.jsx [hunk 6]
```diff
@@ -11,22 +12,51 @@ const VideoStateProvider = ({ children }) => {
...
     }
   };
 
   const [videoState, videoStateDispatch] = useReducer(
-    video_state_reducer,
+    videoStateReducer,
     initialVideoState
   );
-
+  console.log(videoState);
```
Comment (Ankur9669) [inline src/context/VideoStateContext.jsx:55]: Same

File: src/utils/card.util.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+export const isVideoLiked = ({ _id, likedList }) => {
+    if (likedList.length === 0) return false;
+    return likedList.some(video => video._id === _id);
+}
+export const isVideoSaved = ({ _id, playLists }) => {
+    if (playLists.lenght === 0) return false;
```
Comment (Ankur9669) [inline src/utils/card.util.js:6]: fix: length spelling is wrong

**Review by Ankur9669 — APPROVED** [review]
The code looks awesome, i am suggesting some UI changes on which you can work

image

These chips could show the active filter

image

On playing the video it is showing empty screen

image

There Could be some margin beetween text and cards

File: src/context/VideoFilterContext.jsx (hunk 4 above)
Comment (bhtibrewal) [reply src/context/VideoFilterContext.jsx:9]: yes realised it later, changed

File: src/context/VideoStateContext.jsx (hunk 5 above)
Comment (bhtibrewal) [reply src/context/VideoStateContext.jsx:39]: yes done

File: src/context/VideoStateContext.jsx (hunk 6 above)
Comment (bhtibrewal) [reply src/context/VideoStateContext.jsx:55]: done

File: src/utils/card.util.js (hunk 7 above)
Comment (bhtibrewal) [reply src/utils/card.util.js:6]: thank you, for funding this bug

File: src/components/card/BasicCard.jsx (hunk 3 above)
Comment (bhtibrewal) [reply src/components/card/BasicCard.jsx:76]: yes, will do

File: src/context/VideoFilterContext.jsx (hunk 2 above)
Comment (bhtibrewal) [reply src/context/VideoFilterContext.jsx:36]: Thanks, i learn't about it one of kent.c's blogs

**Pull request merged**